jupyter-extension/src/plugins/mount/components/Pipeline/hooks/usePipeline.tsx

```diff
@@ -71,18 +70,26 @@ export const usePipeline = (
     setRequirements(ppsContext?.metadata?.config.requirements ?? '');
     setResponseMessage('');
     if (ppsContext?.metadata?.config.input_spec) {
-      try {
-        setInputSpec(YAML.stringify(ppsContext.metadata.config.input_spec));
-      } catch (_e) {
-        setInputSpec('');
-        setErrorMessage('error parsing input spec'); // This error might confuse user.
-      }
+      setInputSpec(ppsContext.metadata.config.input_spec);
     } else {
       setInputSpec('');
     }
     setCurrentNotebook(ppsContext?.notebookModel?.name ?? 'None');
   }, [ppsContext]);
 
+  useEffect(() => {
+    const ppsMetadata: PpsMetadata = {
+      version: PPS_VERSION,
+      config: {
+        pipeline: pipeline,
+        image: imageName,
+        requirements: requirements,
+        input_spec: inputSpec,
+      },
+    };
+    saveNotebookMetaData(ppsMetadata);
+  }, [pipeline, imageName, requirements, inputSpec]);
```

Comment:
this block is now where autosave is implemented (rather than in the various `onChange` blocks in the component). This fixes a bug where `onChange` would trigger both `setInputSpec` and `callSavePipeline` and, because the latter was async, the two would sometimes race, and the pipeline spec would be saved without the most recent edit.

Now, React manages the update propagation, and `callSavePipeline` (which has been moved into the body of this lambda) is only run after `inputSpec` (and other fields) have been updated, eliminating the race. I've tested this locally and the update actually seems faster than before. Previously, there would sometimes be lag between typing and seeing changes to the notebook metadata, and now it seems consistently close to instant.

Reply:
Nice use of `useEffect` here to manage side effects

Reply:
The only downside I see to this approach is that if/when we add a new field to this pipeline view that we would want to save, we need to remember to add it here.

Reply:
Yeah, makes sense. I added a comment to `types.ts` indicating that `usePipeline.tsx` needs to be kept in sync. In practice, hopefully they won’t drift too much because, without editing this block, a new field would never be written (hopefully an obvious bug)